src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -505,28 +505,53 @@ private static short InferTypeArityFromMetadataName(string emittedTypeName, out
                 return 0;
             }
 
-            // Given a name corresponding to <unmangledName>`<arity>,
-            // extract the arity.
-            string stringRepresentingArity = emittedTypeName.Substring(indexOfManglingChar);
-
-            int arity;
-            bool nonNumericCharFound = !int.TryParse(stringRepresentingArity, NumberStyles.None, CultureInfo.InvariantCulture, out arity);
-
-            if (nonNumericCharFound || arity < 0 || arity > short.MaxValue ||
-                stringRepresentingArity != arity.ToString())
+            // Given a name corresponding to <unmangledName>`<arity>, extract the arity.
+            if (!tryScanArity(emittedTypeName[indexOfManglingChar..], out var arity))
             {
                 suffixStartsAt = -1;
                 return 0;
             }
 
             suffixStartsAt = indexOfManglingChar - 1;
-            return (short)arity;
+            return arity;
+
+            static bool tryScanArity(ReadOnlySpan<char> aritySpan, out short arity)
+            {
+                // Common case: aritySpan is 1 digit between 0-9.  No need for any complex parsing logic in that case.
+                if (aritySpan is [var firstChar and >= '0' and <= '9'])
+                {
+                    arity = (short)(firstChar - '0');
+                    return true;
+                }
+
+#if NET
+                bool nonNumericCharFound = !int.TryParse(aritySpan, NumberStyles.None, CultureInfo.InvariantCulture, out var intArity);
+#else
+                bool nonNumericCharFound = !int.TryParse(aritySpan.ToString(), NumberStyles.None, CultureInfo.InvariantCulture, out var intArity);
+#endif
+
+                if (nonNumericCharFound || intArity < 0 || intArity > short.MaxValue ||
+                    !aritySpan.SequenceEqual(intArity.ToString().AsSpan()))
+                {
+                    arity = -1;
+                    return false;
+                }
+
+                arity = (short)intArity;
+                return true;
+            }
         }
 
         internal static string InferTypeArityAndUnmangleMetadataName(string emittedTypeName, out short arity)
+        {
+            var resultMemory = InferTypeArityAndUnmangleMetadataName(emittedTypeName.AsMemory(), out arity);
+            return resultMemory.ToString();
```

Comment:
>resultMemory.ToString() [](http://example.com/codeflow?start=19&length=23)

Perhaps should return the original string when possible.

Reply:
this will actually do that if the memory does nothing more than wrap a string.  You will get the same (reference equality) string instance back in that case.

Reply:
> this will actually do that if the memory does nothing more than wrap a string.  You will get the same (reference equality) string instance back in that case.

This is not what the doc says ([https://learn.microsoft.com/en-us/dotnet/api/system.readonlymemory-1.tostring?view=net-7.0](https://learn.microsoft.com/en-us/dotnet/api/system.readonlymemory-1.tostring?view=net-7.0)).
Here is a quote from the docs: "For ReadOnlyMemory<T>, returns **a new instance of string** that represents the characters pointed to by the memory"

Reply:
Soundsl ike teh docs are incorrect.  @stephentoub ?

ToString calls this: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ReadOnlyMemory.cs#L144

Because the object is a string in this case, we will call into the `str.Substring(_index, _length)` codepath.  That itself calls into:

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L2070

Which, as the length will be the same, will return `this` as the instance.